libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1588,12 +1595,24 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 			{
 				ArrayUtils(m_context).accessIndex(arrayType, true);
 				if (arrayType.baseType()->isValueType())
-					CompilerUtils(m_context).loadFromMemoryDynamic(
-						*arrayType.baseType(),
-						true,
-						!arrayType.isByteArray(),
-						false
-					);
+				{
+					if (
+						!arrayType.isByteArray() &&
+						arrayType.baseType()->storageBytes() != 32 &&
+						m_context.experimentalFeatureActive(ExperimentalFeature::ABIEncoderV2)
+					)
+					{
+						m_context << u256(32);
+						CompilerUtils(m_context).abiDecodeV2({arrayType.baseType()}, false);
```

Comment:
This will still have quite some overhead in superfluous bounds checking now, won't it? I wonder how much that is...
```
function f(uint64[] calldata a) external returns (uint256 sum) {
  for (uint i = 0; i < 32; i++)
    sum += a[i];
}
```
This will now check once during the initial argument decoding that the whole array fits into calldata - but it will then have spurious bounds checks on each element access as well, since it can't inline the ``m_context << u256(32);`` one line above, right?

Reply:
That's right. I would not worry too much about ABIV2 and non-yul code generator, though...